addons/udes_stock/models/stock_move_line.py

```diff
@@ -0,0 +1,162 @@
+# -*- coding: utf-8 -*-
+from odoo import models, fields
+from collections import defaultdict
+from odoo.tools.float_utils import float_compare, float_round
+
+
+class StockMoveLine(models.Model):
+    _inherit = 'stock.move.line'
+
+    @api.onchange('move_id')
+    def _get_pick_type(self):
+        return self.move_id.picking_type_id.id if self.move_id else False
+
+    u_picking_type_id = fields.Many2one(
+        'stock.picking.type', 'Operation Type', default=_get_pick_type
+    )
+
+    def get_lines_todo(self):
+        """ Return the move lines in self that are not completed,
+            i.e., quantity done < quantity todo
+        """
+        return self.filtered(lambda ml: ml.qty_done < ml.product_uom_qty)
+
+    def get_lines_done(self):
+        """ Return the move lines in self that are completed,
+            i.e., quantity done == quantity todo
+            TODO: Decide on definition of Done, qty_done > 0 or qty_done = expected qty
+        """
+        return self.filtered(lambda ml: ml.qty_done == ml.product_uom_qty)
+
+    def move_lines_for_qty(self, quantity, sort=True):
+        """ Return a subset of move lines from self where their sum of quantity
+            to do is equal to parameter quantity.
+            In case that a move line needs to be split, the new move line is
+            also returned (this happens when total quantity in the move lines is
+            greater than quantity parameter).
+            If there is not enough quantity to do in the mov lines,
+            also return the remaining quantity.
+        """
+        new_ml = None
+        if quantity == 0:
+            return result, new_ml, quantity
+        # TODO: instead of sorting + filtered use a manual search to find
+        #       an equal otherwise a greater. If nothing found then use all
+        #       of them.
+
+        if sort:
+            sorted_mls = self.sorted(lambda ml: ml.product_qty, reverse=True)
+            greater_equal_mls = sorted_mls.filtered(lambda ml: ml.product_qty >= quantity)
+            # last one will be at least equal
+            mls = greater_equal_mls[-1] if greater_equal_mls else sorted_mls
+        else:
+            mls = self
+
+        result = self.browse()
+        for ml in mls:
+            if ml.product_qty >= quantity:
+                extra_qty = ml.product_qty - quantity
+                if extra_qty > 0:
+                    new_ml = ml._split(qty=extra_qty)
+                quantity = 0
+            else:
+                quantity -= ml.product_qty
+            result |= ml
+            if quantity == 0:
+                break
+
+        return result, new_ml, quantity
+
+    def get_quants(self):
+        """ Returns the quants related to move lines in self """
+        Quant = self.env['stock.quant']
+
+        quants = Quant.browse()
+        for ml in self:
+            quants |= Quant._gather(
```

Comment:
Should we be calling private methods outside their class? Core Odoo generally doesn't seem to use `_gather` outside the `Quant` class, except in tests.

Reply:
You're right, _gather in core odoo is not used that much but we do override the core _gather in stock_quant, so I guess we must make more use of it.

Reply:
Rewritten the function.